torch_np/_ndarray.py

```diff
@@ -102,6 +114,13 @@ def __str__(self):
 
     ### comparisons ###
     def __eq__(self, other):
+        try:
+            t_other = asarray(other).get
+        except RuntimeError:
+            # Failed to convert other to array: definitely not equal.
+            # TODO: generalize, delegate to ufuncs
+            falsy = torch.full(self.shape, fill_value=False, dtype=bool)
+            return asarray(falsy)
```

Comment:
I don't think this is correct
```python
>>> np.ndarray([5]) == [3, 4, 5]
<stdin>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
False
```

Reply:
Hm. I thought that using `np.ndarray` was merely not recommended as an array constructor, but did not realize that it does behave *this* differently from a regular `np.array`:

```
In [12]: np.ndarray([6]) == [1, 2, 3]
<ipython-input-12-f4f66d3e4ff6>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  np.ndarray([6]) == [1, 2, 3]
Out[12]: False

In [13]: np.array([6]) == [1, 2, 3]                  # np.array, not *nd*array
Out[13]: array([False, False, False])

In [14]: np.array([6]) == 'oops'
<ipython-input-14-0ccc45613206>:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  np.array([6]) == 'oops'
Out[14]: False
```

I think this because the first argument of `np.ndarray` is `shape`: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html

```
In [19]: np.ndarray([5])
Out[19]: 
array([4.65458471e-310, 0.00000000e+000, 4.94065646e-324,             nan,
       6.92900938e-310])
```

So your example is essentially `np.empty(5) == [1, 2, 3]` and that rightly fails, while `np.ndarray([5]) == [1, 2, 3, 4, 5]` gives an array of five `False`s.

Reply:
That said, this specific try-catch intends to catch a more common issue: consider `np.array([3]) == "oops"`. Since we don't have object arrays, `asarray(other)` fails and we need to explicitly emulate what NumPy's FutureWarning promises. Since `other` is not an array-like, out best bet is to treat it as a scalar, which gives a False on a comparison to each element of `self`.

Reply:
About my example, yep, I know it creates a tensor of 5 elements. My point is that the current code never returns `False`, while NumPy may return a straight `False` from `__eq__`. Another example is the `"oops"` you've written there, which also returns `False`, not a tensor full of `False`s.

Reply:
Yes, the current version here follows the "but in the future will perform elementwise comparison" from the deprecation message. With the "NumPy 2.0" discussion in the numpy ML, it looks like the timescale for the future is about a year.

Reply:
Ah, right, fair.